src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,129 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size;
+    private T[] arrList;
+    private T value;
+
+    public ArrayList() {
+        arrList = (T[]) new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+
+    public void isIndexOutOfBounds(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is not valid");
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        for (int i = 0; i <= arrList.length; i++) {
+            if (i == arrList.length) {
+                grow();
+            }
+            if (i == size) {
+                arrList[i] = value;
+                size++;
+                break;
+            }
+        }
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is not valid");
+        }
+        if (size == arrList.length) {
+            grow();
+        }
+        for (int i = size; i > 0; i--) {
+            if (index < i) {
+                arrList[i] = arrList[i - 1];
+            }
+        }
+        arrList[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = size; i < size + list.size(); i++) {
+            if (size + list.size() > arrList.length) {
+                grow();
+            }
+            arrList[i] = list.get(i - size);
+        }
+        size = size + list.size();
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexOutOfBounds(index);
+        return arrList[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        isIndexOutOfBounds(index);
+        arrList[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isIndexOutOfBounds(index);
+        for (int i = 0; i < size; i++) {
```

Comment:
[Common mistakes](https://mate-academy.github.io/jv-program-common-mistakes/java-web/web-security/java-web-security)
Use [System.arraycopy()](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#arraycopy-java.lang.Object-int-java.lang.Object-int-int-) to move your array elements.